python/ql/src/Functions/IterReturnsNonSelf.ql

```diff
@@ -4,27 +4,94 @@
  * @kind problem
  * @tags reliability
  *       correctness
+ *       quality
  * @problem.severity error
  * @sub-severity low
  * @precision high
  * @id py/iter-returns-non-self
  */
 
 import python
+import semmle.python.ApiGraphs
 
-Function iter_method(ClassValue t) { result = t.lookup("__iter__").(FunctionValue).getScope() }
+/** Holds if `f` is a method of the class `c`. */
+private predicate methodOfClass(Function f, Class c) {
+  exists(FunctionDef d | d.getDefinedFunction() = f and d.getScope() = c)
+}
+
+/** Gets the __iter__ method of `c`. */
+Function iterMethod(Class c) { methodOfClass(result, c) and result.getName() = "__iter__" }
+
+/** Gets the `__next__` method of `c`. */
+Function nextMethod(Class c) { methodOfClass(result, c) and result.getName() = "__next__" }
+
+/** Holds if `var` is a variable referring to the `self` parameter of `f`. */
+predicate isSelfVar(Function f, Name var) { var.getVariable() = f.getArg(0).(Name).getVariable() }
 
-predicate is_self(Name value, Function f) { value.getVariable() = f.getArg(0).(Name).getVariable() }
+/** Holds if `e` is an expression that an iter function `f` should return. */
+predicate isGoodReturn(Function f, Expr e) {
+  isSelfVar(f, e)
+  or
+  exists(DataFlow::CallCfgNode call, DataFlow::AttrRead read, DataFlow::Node selfNode |
+    e = call.asExpr()
+  |
+    call = API::builtin("iter").getACall() and
+    call.getArg(0) = read and
+    read.accesses(selfNode, "__next__") and
+    isSelfVar(f, selfNode.asExpr()) and
+    call.getArg(1).asExpr() instanceof None
+  )
+}
 
-predicate returns_non_self(Function f) {
+/** Holds if the iter method `f` does not return `self` or an equivalent. */
+predicate returnsNonSelf(Function f) {
   exists(f.getFallthroughNode())
   or
-  exists(Return r | r.getScope() = f and not is_self(r.getValue(), f))
+  exists(Return r | r.getScope() = f and not isGoodReturn(f, r.getValue()))
   or
   exists(Return r | r.getScope() = f and not exists(r.getValue()))
```

Comment:
I think the second disjunct subsumes the third one now (or rather, the third one implies the second) because if `r.getValue()` doesn't exist, then surely `isGoodReturn(f, r.getValue())` can't hold either. So there's really no reason for the third disjunct to be there.

(Actually, I guess this was true of the original code as well.)